crates/bevy_asset/src/asset_changed.rs

```diff
@@ -0,0 +1,475 @@
+};
+use bevy_utils::{get_short_name, HashMap};
+
+use crate::{Asset, AssetId, Handle};
+
+#[doc(hidden)]
+#[derive(Resource)]
+pub struct AssetChanges<A: Asset> {
+    change_ticks: HashMap<AssetId<A>, Tick>,
+    last_change_tick: Tick,
+}
+
+impl<A: Asset> AssetChanges<A> {
+    pub(crate) fn insert(&mut self, asset_id: AssetId<A>, tick: Tick) {
+        self.last_change_tick = tick;
+        self.change_ticks.insert(asset_id, tick);
+    }
+    pub(crate) fn remove(&mut self, asset_id: &AssetId<A>) {
+        self.change_ticks.remove(asset_id);
+    }
+}
+impl<A: Asset> Default for AssetChanges<A> {
+    fn default() -> Self {
+        Self {
+            change_ticks: Default::default(),
+            last_change_tick: Tick::new(0),
+        }
+    }
+}
+
+struct AssetChangeCheck<'w, A: Asset> {
+    handle_change_map: &'w HashMap<AssetId<A>, Tick>,
+    last_run: Tick,
+    this_run: Tick,
+}
+impl<A: Asset> Clone for AssetChangeCheck<'_, A> {
+    fn clone(&self) -> Self {
+        *self
+    }
+}
+impl<A: Asset> Copy for AssetChangeCheck<'_, A> {}
+impl<'w, A: Asset> AssetChangeCheck<'w, A> {
+    fn new(changes: &'w AssetChanges<A>, last_run: Tick, this_run: Tick) -> Self {
+        Self {
+            handle_change_map: &changes.change_ticks,
+            last_run,
+            this_run,
+        }
+    }
+    // TODO(perf): some sort of caching? Each check has two levels of indirection,
+    // which is not optimal.
+    fn has_changed(&self, handle: &Handle<A>) -> bool {
+        let is_newer = |tick: &Tick| tick.is_newer_than(self.last_run, self.this_run);
+        let id = &handle.id();
+
+        self.handle_change_map.get(id).is_some_and(is_newer)
+    }
+}
+
+/// A shortcut for the commonly used `Or<(Changed<Handle<A>>, AssetChanged<A>)>`
+/// query filter.
+///
+/// If you want to react to changes to `Handle<A>`, you need to:
+/// - Check if the `Handle<A>` was changed through a `Query<&mut Handle<A>>`.
+/// - Check if the `A` in `Assets<A>` pointed by `Handle<A>` was changed through an
+///   [`Assets<A>::get_mut`].
+///
+/// To properly handle both cases, you need to combine the `Changed` and `AssetChanged`
+/// filters. This query filter is exactly this.
+///
+/// [`Assets<A>::get_mut`]: crate::Assets::get_mut
+pub type AssetOrHandleChanged<A> = Or<(Changed<Handle<A>>, AssetChanged<A>)>;
+
+/// Filter that selects entities with a `Handle<A>` for an asset that changed
+/// after the system last ran.
+///
+/// Unlike `Changed<Handle<A>>`, this is true whenever the asset for the `Handle<A>`
+/// in `ResMut<Assets<A>>` changed. For example, when a mesh changed through the
+/// [`Assets<Mesh>::get_mut`] method, `AssetChanged<Mesh>` will iterate over all
+/// entities with the `Handle<Mesh>` for that mesh. Meanwhile, `Changed<Handle<Mesh>>`
+/// will iterate over no entities.
+///
+/// You probably need to check for both asset and handle changes. If this is the case,
+/// use [`AssetOrHandleChanged`] instead.
```

Comment:
I still don't understand what's a "handle change" so unlikely I'm led to think I need this :) Isn't the common case the case where the asset itself changes?